akka-http-caching/src/test/scala/akka/http/caching/ExpiringLfuCacheSpec.scala

```diff
@@ -0,0 +1,135 @@
+package akka.http.caching
+
+import java.util.Random
+import java.util.concurrent.{ CountDownLatch, Executor }
+
+import akka.actor.ActorSystem
+import akka.testkit.TestKit
+import com.github.benmanes.caffeine.cache.AsyncCacheLoader
+import org.scalatest.{ BeforeAndAfterAll, Matchers, WordSpec }
+
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future, Promise }
+
+class ExpiringLfuCacheSpec extends WordSpec with Matchers with BeforeAndAfterAll {
+  implicit val system = ActorSystem()
+  import system.dispatcher
+
+  implicit def stringLoader = (k: Any) ⇒ Future.successful("")
+  implicit def intLoader = (k: Any) ⇒ Future.successful(0)
+
+  "An LfuCache" should {
+    "be initially empty" in {
+      val cache = lfuCache[String]()
+      cache.store.synchronous.asMap().size should be(0)
+      cache.size should be(0)
+      cache.keys should be(Set())
+    }
+    "store uncached values" in {
+      val cache = lfuCache[String]()
+      Await.result(cache(1).apply("A"), 3.seconds) should be("A")
+      cache.size should be(1)
+      cache.keys should be(Set(1))
+    }
+    "return stored values upon cache hit on existing values" in {
+      val cache = lfuCache[String]()
+      Await.result(cache(1)("A"), 3.seconds) should be("A")
+      cache.size should be(1)
+    }
+    "return Futures on uncached values during evaluation and replace these with the value afterwards" in {
+      val cache = lfuCache[String]()
+      val latch = new CountDownLatch(1)
+      val future1 = cache(1) { (promise: Promise[String]) ⇒
+        Future {
+          latch.await()
+          promise.success("A")
+        }
+      }
+      val future2 = cache(1)("")
+      Thread.sleep(50)
+      cache.store.get(1).getNumberOfDependents should be(2)
+
+      latch.countDown()
+      Await.result(future1, 3.seconds) should be("A")
+      Await.result(future2, 3.seconds) should be("A")
+      cache.size should be(1)
+    }
+    "properly limit capacity" in {
+      val cache = lfuCache[String](maxCapacity = 3)
+      Await.result(cache(1)("A"), 3.seconds) should be("A")
+      Await.result(cache(2)(Future.successful("B")), 3.seconds) should be("B")
+      Await.result(cache(3)("C"), 3.seconds) should be("C")
+      cache(4)("D")
+      Thread.sleep(50)
+      cache.size should be(3)
+    }
+    "expire old entries" in {
+      val cache = lfuCache[String](timeToLive = 75 millis span)
+      Await.result(cache(1)("A"), 3.seconds) should be("A")
+      Await.result(cache(2)("B"), 3.seconds) should be("B")
+      Thread.sleep(50)
```

Comment:
After repeated runs it still occassionally fails to evict an item - even with calls to `.synchronous().cleanUp()` and after initialisation with an `Executor`.
It would be nice to get this working reliably, though it's tempting to remove these tests since they're now effectively testing a thin wrapper around caffeine, which I'm confident is well tested :)

Here's how I'm currently testing it - this often passes many times out of 50 before failing:
```
for (a ← 0 to 50) {
  val cache = lfuCache[String](timeToLive = 75 millis span)

  Await.result(cache(1)("A"), 3.seconds) should be("A")
  ticker.advance(50 millis span)

  Await.result(cache(2)("B"), 3.seconds) should be("B")
  cache.size should be(2)  //always passes

  ticker.advance(50 millis span)

  eventually {  //repeatedly check if the item has been evicted
    cache.store.synchronous().cleanUp()
    Thread.sleep(50)

    //If this fails once then it never passes, no matter how many times it runs
    cache.get(1) should be(None)
    cache.size should be(1)
  }
}
```